### PR TITLE
Upgrade Devdocs to Ruby 3.1.2

### PR DESCRIPTION
**Pull request by oscillot**
This change:

- Upgrades the ruby version from 2.7.6 to the current latest stable 3.1.2
- Refreshes gem versions and updates code where needed
  - Some of the most common changes have to do with the changes to arguments introduced in 2.7 and finalized in 3.0. Mostly this was cleaning up some uses of `*` and `**` splats, and one or two uses of `&` that needed to be added
- The browser gem no longer ships a class method (:modern?), I have done my best to recreate the previous behavior faithfully in `lib/app.rb` though I wonder if we even need these checks any longer?

- [X] Tested the changes locally to ensure:
  - [X] All scrapers still work without errors
  - [X] The scraped documentation still looks consistent
- [X] Built the docker image successfully 
  - [X] Tested the application running in the container successfully
- [X] Built the alpine-flavored docker image and ran devdocs successfully
  - [X] Tested the application running in the container successfully
- [X] All tests are passing

### CODE REVIEW

**Comment by oscillot** [comment]
Still kicking the tires on this, need to go and do my day job for a while.

**Review by ojeytonwilliams — COMMENTED** [review]
I'm not particularly great at Ruby, but I didn't see anything worrying in the code. The only things I found were some warnings in the console:

```
Calling `DidYouMean::SPELL_CHECKERS.merge!(error_name => spell_checker)' has been deprecated. Please call `DidYouMean.correct_error(error_name, spell_checker)' instead.
/home/oliver/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/sprockets-3.7.2/lib/sprockets/erb_processor.rb:21: warning: Passing safe_level with the 2nd argument of ERB.new is deprecated. Do not use it, and specify other arguments as keyword arguments.
```

Neither of those are present when using ruby 2.

That said, they're warnings and the app seemed to function just fine.

**Comment by oscillot** [comment]
> I'm not particularly great at Ruby, but I didn't see anything worrying in the code. The only things I found were some warnings in the console:
> 
> ```
> Calling `DidYouMean::SPELL_CHECKERS.merge!(error_name => spell_checker)' has been deprecated. Please call `DidYouMean.correct_error(error_name, spell_checker)' instead.
> /home/oliver/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/sprockets-3.7.2/lib/sprockets/erb_processor.rb:21: warning: Passing safe_level with the 2nd argument of ERB.new is deprecated. Do not use it, and specify other arguments as keyword arguments.
> ```
> 
> Neither of those are present when using ruby 2.
> 
> That said, they're warnings and the app seemed to function just fine.

I did look into these. These are actually coming from Thor. I've bumped us to the latest which is 1.2.1 where the issue still exists: https://github.com/rails/thor/blob/v1.2.1/lib/thor/error.rb#L109-L112

There is some movement on main https://github.com/rails/thor/blob/main/lib/thor/error.rb but I think this is on that group of maintainers to resolve. We can pull in the new version whenever they cut it, I'm sure the rails team is aware of it. Push comes to shove we can convert these to rake tasks, it wouldn't be a huge lift and we'd already be doing another ruby upgrade to get ourselves into that situation. `DidYouMean` is built-in to the language now so the deprecation will happen in lockstep with a future version. 

For now it looks like we are fine until ruby 3.3: https://github.com/ruby/ruby/blob/f9712b029bb6c34a2cf4131f8a54328740fd149f/lib/did_you_mean.rb#L116-L133

**Comment by oscillot** [comment]
Okay! The docker builds look alright. When building the docker images there's one more deprecation warning that comes up:

```
/usr/local/bundle/gems/sprockets-3.7.2/lib/sprockets/erb_processor.rb:21: warning: Passing safe_level with the 2nd argument of ERB.new is deprecated. Do not use it, and specify other arguments as keyword arguments.
/usr/local/bundle/gems/sprockets-3.7.2/lib/sprockets/erb_processor.rb:21: warning: Passing trim_mode with the 3rd argument of ERB.new is deprecated. Use keyword argument like ERB.new(str, trim_mode: ...) instead.
```

This is coming from sprockets-sass which is EOL. Other than that the docker builds look good to me, although I am still new to this codebase. The recommended replacement is dart-sass but that seems well out of scope for this set of changes.

There is one more thing that I want to make sure is sorted. Whoever deploys this in heroku will need to set up the buildpack for 3.1.2, it's one of the 3 currently supported ruby buildpacks. Failure to do this could would almost certainly prevent the app from starting and we'd need to roll back to the 2.7.6 version

I did add one more change that is for `asdf` users, pretty straight-forward, just something I missed on the first pass. Let me know if there are any questions or anything else I can help with.

File: docs/maintainers.md [hunk 1]
```diff
@@ -125,6 +126,7 @@ The following people (used to) maintain DevDocs:
 - [Jed Fox](https://github.com/j-f1)
 - [Mrugesh Mohapatra](https://github.com/raisedadead)
 - [Oliver Eyton-Williams](https://github.com/ojeytonwilliams)
+- [Paul Sernatinger](https://github.com/oscillot)
```
Comment (ojeytonwilliams) [inline docs/maintainers.md:129]: Hey @oscillot, if you'd like to become a maintainer, that's awesome. It's more of a long term thing, though, so if you want to stick around and continue finding ways to contribute, please do.

Basically, if you maintain DevDocs for a bit we can make it official!

File: lib/app.rb [hunk 2]
```diff
@@ -238,7 +238,22 @@ def redirect_via_js(path)
     end
 
     def supports_js_redirection?
-      browser.modern? && !memoized_cookies.empty?
+      modern_browser?(browser) && !memoized_cookies.empty?
+    end
```
Comment (simon04) [inline lib/app.rb:242]: This check can be dropped completely, as "DevDocs being a developer tool, the browser requirements are high" (citation from README).

File: Gemfile.lock [hunk 3]
```diff
@@ -176,7 +179,7 @@ DEPENDENCIES
   yajl-ruby
 
 RUBY VERSION
-   ruby 2.7.3p183
+   ruby 3.1.2p20
 
 BUNDLED WITH
    2.1.4
```
Comment (simon04) [inline Gemfile.lock:185]:
```suggestion
BUNDLED WITH
   2.3.11
```

Fixes "Calling `DidYouMean::SPELL_CHECKERS.merge!(error_name => spell_checker)' has been deprecated. Please call `DidYouMean.correct_error(error_name, spell_checker)' instead."

See `https://github.com/github/explore/pull/3020`

**Review by simon04 — COMMENTED** [review]
Thank you for this great PR. I've performed some tests: 

```
> bundle exec thor docs:download vite
(1/1) Vite OK
Done
```

```
> bundle exec thor docs:generate --force --package vite
OK
```

```
> bundle exec thor docs:generate --force --package react
[#####                                                                                                                                                                             ] [ 2/69] [  2.90%] [00:00] [00:30] [ 2.18/s]exception when reading /Users/simon/src/devdocs/.image_optim.yml: Tried to load unspecified class: Range
svgo worker: `svgo` not found; please provide proper binary or disable this worker (--no-svgo argument or `:svgo => false` through options)
ERROR: https://reactjs.org/static/89baed0a6540f29e954065ce04661048/13ae7/cdn-cors-header.png                                                                                                                                    
  NameError: uninitialized constant Docs::ImagesFilter::Base64
    
                image = Base64.strict_encode64(image)
                        ^^^^^^

  /Users/simon/src/devdocs/lib/docs/filters/core/images.rb:69:in `block (2 levels) in call'
  /Users/simon/src/devdocs/lib/docs/core/request.rb:27:in `block in run'
  /Users/simon/src/devdocs/lib/docs/core/instrumentable.rb:15:in `instrument'
  /Users/simon/src/devdocs/lib/docs/core/request.rb:26:in `run'
  /Users/simon/src/devdocs/lib/docs/core/request.rb:13:in `run'
  /Users/simon/src/devdocs/lib/docs/filters/core/images.rb:38:in `block in call'
  /Users/simon/src/devdocs/lib/docs/filters/core/images.rb:23:in `call'
  /Users/simon/src/devdocs/lib/docs/core/instrumentable.rb:15:in `instrument'
  /Users/simon/src/devdocs/lib/docs/core/instrumentable.rb:15:in `instrument'
  /Users/simon/src/devdocs/lib/docs/core/scraper.rb:179:in `process_response'
  /Users/simon/src/devdocs/lib/docs/core/scraper.rb:160:in `block in handle_response'
  /Users/simon/src/devdocs/lib/docs/core/instrumentable.rb:15:in `instrument'
  /Users/simon/src/devdocs/lib/docs/core/scraper.rb:159:in `handle_response'
  /Users/simon/src/devdocs/lib/docs/core/scraper.rb:77:in `block in build_pages'
  /Users/simon/src/devdocs/lib/docs/core/requester.rb:59:in `block (2 levels) in handle_response'
  /Users/simon/src/devdocs/lib/docs/core/requester.rb:58:in `each'
  /Users/simon/src/devdocs/lib/docs/core/requester.rb:58:in `block in handle_response'
  /Users/simon/src/devdocs/lib/docs/core/instrumentable.rb:15:in `instrument'
  /Users/simon/src/devdocs/lib/docs/core/requester.rb:57:in `handle_response'
  /Users/simon/src/devdocs/lib/docs/core/requester.rb:18:in `run'
  /Users/simon/src/devdocs/lib/docs/core/scrapers/url_scraper.rb:38:in `request_all'
  /Users/simon/src/devdocs/lib/docs/core/scraper.rb:76:in `build_pages'
  /Users/simon/src/devdocs/lib/docs/core/doc.rb:115:in `block in store_pages'
  /Users/simon/src/devdocs/lib/docs/storage/abstract_store.rb:87:in `block (2 levels) in replace'
  /Users/simon/src/devdocs/lib/docs/storage/abstract_store.rb:182:in `track_touched'
  /Users/simon/src/devdocs/lib/docs/storage/abstract_store.rb:87:in `block in replace'
  /Users/simon/src/devdocs/lib/docs/storage/abstract_store.rb:170:in `lock'
  /Users/simon/src/devdocs/lib/docs/storage/abstract_store.rb:87:in `replace'
  /Users/simon/src/devdocs/lib/docs/storage/abstract_store.rb:85:in `block in replace'
  /Users/simon/src/devdocs/lib/docs/storage/abstract_store.rb:144:in `open_yield_close'
  /Users/simon/src/devdocs/lib/docs/storage/abstract_store.rb:30:in `open'
  /Users/simon/src/devdocs/lib/docs/storage/abstract_store.rb:85:in `replace'
  /Users/simon/src/devdocs/lib/docs/core/doc.rb:114:in `store_pages'
  /Users/simon/src/devdocs/lib/docs.rb:100:in `generate'
  /Users/simon/src/devdocs/lib/tasks/docs.thor:303:in `generate_doc'
  /Users/simon/src/devdocs/lib/tasks/docs.thor:105:in `generate'
```

**Comment by simon04** [comment]
> There is one more thing that I want to make sure is sorted. Whoever deploys this in heroku will need to set up the buildpack for 3.1.2, it's one of the 3 currently supported ruby buildpacks. Failure to do this could would almost certainly prevent the app from starting and we'd need to roll back to the 2.7.6 version

@ojeytonwilliams, do you know how/where to configure the buildpack?

**Comment by ojeytonwilliams** [comment]
> do you know how/where to configure the buildpack?

It should pick up the ruby version from the Gemfile and use it automatically.  The only caveat is that it needs to be on the heroku-22 stack. So, when you're happy with this PR let me know and I'll test deploying it to that stack.

If all goes well, I'll upgrade the main dyno and merge this.

**Comment by oscillot** [comment]
@simon04 I will take a look at that error, It looks like it just needs to be `::Base64`. Less sure about svgo, but I can dig into that

File: docs/maintainers.md (hunk 1 above)
Comment (oscillot) [reply docs/maintainers.md:129]: ++ Understood, I can remove that for now

File: lib/app.rb (hunk 2 above)
Comment (oscillot) [reply lib/app.rb:242]: @simon04 Double checking: we don't need just the `modern_browser?` check within this method or are you saying that the whole `supports_js_redirection?` method can be dropped?

**Comment by oscillot** [comment]
Okay, I've got everything sorted **except** the `image_optim` thing. It has something to do with psych (maybe via yajl?) and using safe_load. I'll try to sort it out tomorrow with a fresh set of eyes. Really appreciate the thorough review and patience while I work through the final bits!

**Comment by oscillot** [comment]
Okay, we definitely have a blocker with `image_optim`. I have [opened an issue against image_optim](https://github.com/toy/image_optim/issues/203) this morning and will see how they reply.

If we want to get unblocked here we could introduce a refinement to work around this issue, but I think it's fine to wait until things are fixed in image_optim. According to the Heroku docs [we should have until 04/2025 to get off of heroku-20](https://devcenter.heroku.com/articles/heroku-20-stack) so the question is really one of patience and priorities.

@simon04 @ojeytonwilliams Let me know what you folks think and in the meantime I'll do my best to shephard this issue along to a resolution

**Comment by oscillot** [comment]
I included this in the issue against image_optim, [this blog post](https://www.ctrl.blog/entry/ruby-psych4.html) did a good job at sending me in the right direction and then I was able to [reproduce the issue in isolation](https://github.com/oscillot/image_optim_ruby_3.1_psych_4_repro) to include with the issue

**Comment by ojeytonwilliams** [comment]
> I think it's fine to wait until things are fixed in image_optim

Yeah, that makes sense. As you say, there's no great rush, so we can take the time to do this properly.

**Comment by oscillot** [comment]
Okay, I've [PRed against image_optim](https://github.com/toy/image_optim/pull/204) this morning, We'll see how this goes.

**Review by simon04 — APPROVED** [review]
https://github.com/toy/image_optim/pull/204 has been merged and released as [0.31.3](https://rubygems.org/gems/image_optim/versions/0.31.3), thanks @toy and @oscillot 

I've amended this merge request to feature Ruby 3.2.1

I've tested generating/scraping a few docs and running the local development server successfully.

I'd like to give Ruby 3.2.1 a try. What do you think @ojeytonwilliams? Do you want to take any preparatory steps?

**Comment by ojeytonwilliams** [comment]
Hey @simon04 I ran into some issues when trying to deploy this. `rake assets:precompile` was failing when it tried to call `open`. I've pushed what seems to be a fix, but could you confirm I've not done something silly?

The current Heroku stack can handle this Ruby version. So, if this change is okay, then we can merge this as is. 

I'll update the stack after that, assuming all goes well.

**Comment by simon04** [comment]
Hey Oliver, thank you for your preparatory tests and for addressing the
`open` problem. Your fix looks okay. :)

Feel free to proceed. I will be around later today, take a look and try to
build some new docs. Right now, I'm chilling on a 2600m asl mountain in
Austria. 😎

**Pull request merged**